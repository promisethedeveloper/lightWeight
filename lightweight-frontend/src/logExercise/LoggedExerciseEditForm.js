import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { useParams } from "react-router-dom";
import Alert from "../common/Alert";
import LightWeightAPI from "../api/api";

const LoggedExerciseEditForm = () => {
	const { id, weight, unit, no_of_sets, no_of_reps } = useParams();

	const INITIALSTATE = {
		weight,
		unit,
		no_of_sets,
		no_of_reps,
	};

	const history = useHistory();
	const [formData, setFormData] = useState(INITIALSTATE);
	const [formErrors, setFormErrors] = useState([]);

	/**
	 * Handle form submit:
	 *
	 * Calls login function prop and if successful, redirect to "/workouts"
	 */

	const handleSubmit = async (e) => {
		e.preventDefault();
		try {
			let { weight, unit, no_of_sets, no_of_reps } = formData;
			let data = {
				weight: +weight,
				unit,
				no_of_sets: +no_of_sets,
				no_of_reps: +no_of_reps,
			};
			await LightWeightAPI.editLoggedExercise(id, data);
			history.push("/workouts");
			setFormData(INITIALSTATE);
		} catch (error) {
			setFormErrors(error);
		}
	};

	const handleChange = (e) => {
		const { name, value } = e.target;
		setFormData((data) => ({
			...data,
			[name]: value,
		}));
	};

	return (
		<div className="LoginForm">
			<div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
				<h3 className="mb-3">Edit exercise</h3>
				<div className="card">
					<div className="card-body">
						<form onSubmit={handleSubmit}>
							<div className="form-group">
								<label htmlFor="weight">Weight</label>
								<input
									id="weight"
									type="number"
									name="weight"
									className="form-control"
									value={formData.weight}
									onChange={handleChange}
									placeholder="ex. 100"
									required
								/>
							</div>
							<div className="form-group">
								<label htmlFor="unit">Unit</label>
								<input
									id="unit"
									type="text"
									name="unit"
									className="form-control"
									value={formData.unit}
									onChange={handleChange}
									placeholder="ex. kg or Lbs"
									required
								/>
							</div>
							<div className="form-group">
								<label htmlFor="no_of_sets">No. of sets</label>
								<input
									id="no_of_sets"
									type="number"
									name="noOfSets"
									className="form-control"
									value={formData.no_of_sets}
									onChange={handleChange}
									placeholder="ex. 5"
									required
								/>
							</div>
							<div className="form-group">
								<label htmlFor="no_of_reps">No. of Reps.</label>
								<input
									id="no_of_reps"
									type="number"
									name="noOfReps"
									className="form-control"
									value={formData.no_of_reps}
									onChange={handleChange}
									placeholder="ex. 10"
									required
								/>
							</div>
							{formErrors.length ? (
								<Alert type="danger" messages={formErrors} />
							) : null}
							<button
								className="btn float-right font-weight-bold"
								style={{ backgroundColor: "#461111", color: "white" }}
							>
								Save
							</button>
							<a
								className="btn float-right font-weight-bold"
								style={{
									backgroundColor: "#FF1E00",
									color: "white",
									marginRight: "5px",
								}}
								href="/workouts"
								role="button"
							>
								Cancel
							</a>
						</form>
					</div>
				</div>
			</div>
		</div>
	);
};

export default LoggedExerciseEditForm;
