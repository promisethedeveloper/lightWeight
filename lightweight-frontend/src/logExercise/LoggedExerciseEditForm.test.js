import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import LoggedExerciseEditForm from "./LoggedExerciseEditForm";

/** Smoke test */
it("renders without crashing", () => {
	render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
});

/** Snapshot test */
it("Should match snapshots", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});

it("renders a 'Loading...' text", () => {
	const { getByText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByText("Edit exercise")).toBeInTheDocument();
});

it("renders form's weight label text", () => {
	const { getByLabelText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByLabelText("Weight")).toBeInTheDocument();
});

it("renders form's unit label text", () => {
	const { getByLabelText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByLabelText("Unit")).toBeInTheDocument();
});

it("renders form's no. of sets label text", () => {
	const { getByLabelText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByLabelText("No. of sets")).toBeInTheDocument();
});

it("renders form's no. of reps label text", () => {
	const { getByLabelText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByLabelText("No. of Reps.")).toBeInTheDocument();
});

it("renders form's no. of reps label text", () => {
	const { getByText } = render(
		<MemoryRouter>
			<LoggedExerciseEditForm />
		</MemoryRouter>
	);
	expect(getByText("Save")).toBeInTheDocument();
});
