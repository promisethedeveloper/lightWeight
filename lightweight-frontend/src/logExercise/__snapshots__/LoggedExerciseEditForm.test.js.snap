// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshots 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h3
        class="mb-3"
      >
        Edit exercise
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="weight"
              >
                Weight
              </label>
              <input
                class="form-control"
                id="weight"
                name="weight"
                placeholder="ex. 100"
                required=""
                type="number"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="unit"
              >
                Unit
              </label>
              <input
                class="form-control"
                id="unit"
                name="unit"
                placeholder="ex. kg or Lbs"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="no_of_sets"
              >
                No. of sets
              </label>
              <input
                class="form-control"
                id="no_of_sets"
                name="noOfSets"
                placeholder="ex. 5"
                required=""
                type="number"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="no_of_reps"
              >
                No. of Reps.
              </label>
              <input
                class="form-control"
                id="no_of_reps"
                name="noOfReps"
                placeholder="ex. 10"
                required=""
                type="number"
                value=""
              />
            </div>
            <button
              class="btn float-right font-weight-bold"
              style="background-color: rgb(70, 17, 17); color: white;"
            >
              Save
            </button>
            <a
              class="btn float-right font-weight-bold"
              href="/workouts"
              role="button"
              style="background-color: rgb(255, 30, 0); color: white; margin-right: 5px;"
            >
              Cancel
            </a>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
